### ISSUE TITLE
Request for Better Type Definitions and Interfaces

### ISSUE THREAD

**Issue by roiiiu** [greensock/GSAP#533]
I have found gsap to be the most convenient animation library I have used so far. However, its TypeScript support leaves room for improvement. Due to the lack of precise type definitions, I often find myself having to refer to the official documentation more frequently than I would like. 
For instance, I wonder if it would be beneficial to add literal property union types for the ease attribute, instead of relying on strings. This would not only make the code more self-explanatory and easier to maintain, but also boost productivity and confidence in designing complex animation sequences.
<img width="404" alt="image" src="https://user-images.githubusercontent.com/11494827/232563180-fbf644b3-259d-4bb4-9edd-ab8e9f9cc0f1.png">
<img width="542" alt="image" src="https://user-images.githubusercontent.com/11494827/232564337-ea5f21c1-d9b0-44e0-b979-725d7fe91757.png">
<img width="370" alt="image" src="https://user-images.githubusercontent.com/11494827/232565387-34ac6df2-5060-4bc6-add9-0c54b03e9dac.png">

**Comment by jackdoyle** [comment]
Thanks for the suggestion, @roiiiu 

The `ease` must accommodate ANY string because you can create a CustomEase, CustomBounce, CustomWiggle, or even call `gsap.registerEase()` and assign an arbitrary string as the name. So do you think it'd still be helpful to do something like this? 
```javascript
type EaseName = "none"
            | "power1" | "power1.in" | "power1.out" | "power1.inOut"
            | "power2" | "power2.in" | "power2.out" | "power2.inOut"
            | "power3" | "power3.in" | "power3.out" | "power3.inOut"
            | "power4" | "power4.in" | "power4.out" | "power4.inOut"
            | "back" | "back.in" | "back.out" | "back.inOut"
            | "bounce" | "bounce.in" | "bounce.out" | "bounce.inOut"
            | "circ" | "circ.in" | "circ.out" | "circ.inOut"
            | "elastic" | "elastic.in" | "elastic.out" | "elastic.inOut"
            | "expo" | "expo.in" | "expo.out" | "expo.inOut"
            | "sine" | "sine.in" | "sine.out" | "sine.inOut"
            | "rough({template: none, strength: 1, points: 20, taper: none, randomize: true, clamp: false})" | "slow(0.7, 0.7, false)" | "steps(12)" | "expoScale(1, 2)";
 ```
 And then: 
```javascript
ease?: EaseName | EaseFunction | string;
```
?

**Comment by roiiiu** [comment]
> Thanks for the suggestion, @roiiiu 
> 
> The `ease` must accommodate ANY string because you can create a CustomEase, CustomBounce, CustomWiggle, or even call `gsap.registerEase()` and assign an arbitrary string as the name. So do you think it'd still be helpful to do something like this? 
> ```javascript
> type EaseName = "none"
>             | "power1" | "power1.in" | "power1.out" | "power1.inOut"
>             | "power2" | "power2.in" | "power2.out" | "power2.inOut"
>             | "power3" | "power3.in" | "power3.out" | "power3.inOut"
>             | "power4" | "power4.in" | "power4.out" | "power4.inOut"
>             | "back" | "back.in" | "back.out" | "back.inOut"
>             | "bounce" | "bounce.in" | "bounce.out" | "bounce.inOut"
>             | "circ" | "circ.in" | "circ.out" | "circ.inOut"
>             | "elastic" | "elastic.in" | "elastic.out" | "elastic.inOut"
>             | "expo" | "expo.in" | "expo.out" | "expo.inOut"
>             | "sine" | "sine.in" | "sine.out" | "sine.inOut"
>             | "rough({template: none, strength: 1, points: 20, taper: none, randomize: true, clamp: false})" | "slow(0.7, 0.7, false)" | "steps(12)" | "expoScale(1, 2)";
>  ```
>  And then: 
> ```javascript
> ease?: EaseName | EaseFunction | string;
> ```
> ?

I am sorry for my limited knowledge about the library, but I still believe that using string to specify so many property variables is not the best practice. However, the cost of making changes may have some impact on the current users.😌

**Comment by jackdoyle** [comment]
Oh, we will definitely keep the named strings for eases. In the very old days, we had ease objects for all the eases which technically made strict typing easier but it was clunky because it required more imports/exports (file size tradeoff) and it wasn't conducive to custom eases. 

Plus there are over 11 million sites that rely on GSAP and tons of developers are very accustom to using the named strings for easing, so if we suddenly moved away from that, it would be extremely annoying for those devs and would cause breaking changes. I see absolutely no reason to do that, sorry.

However, if there's something we could do in the TypeScript definitions to make your life easier, I'm very open to that. I'm not much of a TypeScript person, so my question to you was more about whether or not that structure I shared above would actually be helpful at all (like for code hinting in your editor). In other words, if the ease accepts an `EaseName` as well as a `string`, does that give you the code hinting you want or is there no benefit whatsoever to having that `EaseName` type since we must allow a type of `string` anyway?

**Comment by charnog** [comment]
For the official playground and VSCode, it doesn't make any difference. A-la `'asdf'` and `'zxcv'` are subtypes of `string`.
<img width="500" alt="image" src="https://user-images.githubusercontent.com/1962140/232879165-28e896d5-4b4a-489e-a639-34c5dd3e21f9.png">
<img width="500" alt="image" src="https://user-images.githubusercontent.com/1962140/232881857-063e78db-fa70-47ae-b093-28d20a9727da.png">

WebStorm understands it, and gives the hint (they have their own tuned implementation of TS language server):
<img width="350" alt="image" src="https://user-images.githubusercontent.com/1962140/232880019-a75cad38-2947-43f4-8693-b4bf4e8db873.png">
<img width="350" alt="image" src="https://user-images.githubusercontent.com/1962140/232879989-04e87408-dd37-4b94-9cb3-ba87ace681ea.png">

I would say "yes, we need it" if it were consistent across all implementations, especially in VSCode, as many folks use it. I prefer to leave it as a `string`, _at least for now_.

By the way, there are a few things to remember; try to read them as one word:
`"powerX" and "back" and "bounce", "circ", "elastic", "expo", "sine".`

Then you can add `.` and change the type via `"in" or "out" or "inOut"`.

**Comment by jackdoyle** [comment]
> I would say "yes, we need it" if it were consistent across all implementations, especially in VSCode, as many folks use it. I prefer to leave it as a `string`, _at least for now_.

I read that a few times and am still fuzzy. You say that my proposed change DOES help in some editors like WebStorm, right? So "yes we need it"...and then you quickly follow that with "I prefer to leave it as a string" which sounds like "no, we don't need it...stick with a generic string as the type". Can you please clarify what you meant? 

If there's benefit for some (and it doesn't hurt anyone), I'm inclined to do it. 

> By the way, there are a few things to remember; try to read them as one word: `"powerX" and "back" and "bounce", "circ", "elastic", "expo", "sine".`
> 
> Then you can add `.` and change the type via `"in" or "out" or "inOut"`.

Again, this left me confused. Can you rephrase and make your suggestion very clear please? Sorry if I'm missing something obvious.

**Comment by charnog** [comment]
Oops, my bad.

The message has the following structure:

1. The first part shows where it works and where it doesn't (without implying whether it should be implemented or not).
2. The second part presents my opinion on whether it should be implemented or not

> If there's benefit for some (and it doesn't hurt anyone), I'm inclined to do it.

If so then I suggest to use [Template Literal Types](https://www.typescriptlang.org/docs/handbook/2/template-literal-types.html)

<img width="527" alt="image" src="https://user-images.githubusercontent.com/1962140/232893920-2580e16b-0bab-40a1-b399-cc364a343668.png">

You can also inline `Dir` to template, if you don't want to create a distinguish type for it.
```ts
type Ease = "none" | EaseName | `${EaseName}.${"in" | "out" | "inOut"}` | string;
```

**Comment by jackdoyle** [comment]
Thanks for the suggestion, @charnog

Do you know of any compatibility issues with using the template literal types? Should this be pretty universally compatible these days?: 

```javascript
  type EaseName = "power1" | "power2" | "power3" | "power4" | "back" | "bounce" | "circ" | "elastic" | "expo" | "sine";

  type EaseDirection = "in" | "out" | "inOut";

  type EaseString = "none" | "rough({template: none, strength: 1, points: 20, taper: none, randomize: true, clamp: false})" | "slow(0.7, 0.7, false)" | "steps(12)" | "expoScale(1, 2)" | EaseName | `${EaseName}.${EaseDirection}` | string;
```
```javascript
ease?: EaseString | EaseFunction;
```

From a quick glance, they were added in TypeScript 4.1 in late 2020. I don't have a good feel for how dangerous it would be to add into our codebase (I don't want to cause problems in legacy projects for example).

**Comment by charnog** [comment]
> I don't have a good feel for how dangerous it would be to add into our codebase (I don't want to cause problems in legacy projects for example).

If so, then yes, you are right, it may cause problems, and it's better to use a 'plain old string union' as mentioned in [this](https://github.com/greensock/GSAP/issues/533#issuecomment-1512535340) message.

In [that](https://github.com/greensock/GSAP/issues/533#issuecomment-1512535340) case, there are no concerns about 'broken types'. Previously, the type was just `string`. If you update the type to be more specific but still include `string` in the type definition, then nothing will actually change. Users who used `'myCustomEaseName'` will still rely on a `string` within the updated type. WebStorm will show the suggestions for the new types (I don't know why VSCode doesn't).

By the way, I noticed that you wrote `"rough({template: none, strength: 1, points: 20, taper: none, randomize: true, clamp: false})" | "slow(0.7, 0.7, false)" | "steps(12)" | "expoScale(1, 2)"`. Unfortunately, it won't be interpolated, and users will have to use it exactly as you wrote in the type. You can simply removed it, `string` part of the updated type will allow users to write whatever they want, but more specific `'power1.in' ...` will allow IDEs to suggest the simple ones.
<img width="734" alt="image" src="https://user-images.githubusercontent.com/1962140/233411258-f4b5e1f8-a5e3-4314-b902-b2ed68aaff0f.png">

It would be possible with template literals though, like this: `slow(${number}, ${number}, ${boolean})`
<img width="488" alt="image" src="https://user-images.githubusercontent.com/1962140/233410915-3e5e113c-a85a-4939-88b8-6c6bd1e0e431.png">
